monai/apps/deepgrow/handler.py

```diff
@@ -0,0 +1,287 @@
+                for i in range(3):
+                    img = self.plot_data[region][i]
+                    img = np.moveaxis(img, -3, -1)
+                    plot_2d_or_3d_image(
+                        img[np.newaxis], epoch, self.writer, 0, self.max_channels, self.max_frames, tags[i]
+                    )
+
+        self.logger.info(
+            "Saved {} Regions {} into Tensorboard at epoch: {}".format(
+                len(self.plot_data), sorted([*self.plot_data]), epoch
+            )
+        )
+        self.writer.flush()
+
+    def write_region_metrics(self, epoch):
+        metric_sum = 0
+        means = {}
+        for region in self.metric_data:
+            metric = self.metric_data[region].mean()
+            self.logger.info(
+                "Epoch[{}] Metrics -- Region: {:0>2d}, {}: {:.4f}".format(epoch, region, self.tag_name, metric)
+            )
+
+            if self.merge_scalar:
+                means["{:0>2d}".format(region)] = metric
+            else:
+                self.writer.add_scalar("{}_{:0>2d}".format(self.tag_name, region), metric, epoch)
+            metric_sum += metric
+
+        if self.merge_scalar:
+            means["avg"] = metric_sum / len(self.metric_data)
+            self.writer.add_scalars("{}_region".format(self.tag_name), means, epoch)
+        elif len(self.metric_data) > 1:
+            metric_avg = metric_sum / len(self.metric_data)
+            self.writer.add_scalar("{}_regions_avg".format(self.tag_name), metric_avg, epoch)
+        self.writer.flush()
+
+    def __call__(self, engine: Engine, action) -> None:
+        total_steps = engine.state.iteration
+        if total_steps < engine.state.epoch_length:
+            total_steps = engine.state.epoch_length * (engine.state.epoch - 1) + total_steps
+
+        if action == "epoch" and not self.fold_size:
+            epoch = engine.state.epoch
+        elif self.fold_size and total_steps % self.fold_size == 0:
+            epoch = int(total_steps / self.fold_size)
+        else:
+            epoch = None
+
+        if epoch:
+            if self.images and epoch % self.image_interval == 0:
+                self.write_images(epoch)
+            if self.add_scalar:
+                self.write_region_metrics(epoch)
+
+        if action == "epoch" or epoch:
+            self.plot_data = {}
+            self.metric_data = {}
+            return
+
+        device = engine.state.device
+        batch_data = engine.state.batch
+        output_data = engine.state.output
+
+        for bidx in range(len(batch_data.get("region", []))):
+            region = batch_data.get("region")[bidx]
+            region = region.item() if torch.is_tensor(region) else region
+
+            if self.images and self.plot_data.get(region) is None:
+                self.plot_data[region] = [
+                    rescale_array(batch_data["image"][bidx][0].detach().cpu().numpy()[np.newaxis], 0, 1),
```

Comment:
why does "image" need this trick while "label" and "pred" does not?

Reply:
removed this code.. not need as part of core functionality for deepgrow